### ISSUE TITLE
Memory bug in validation

### ISSUE THREAD

**Issue by tomkocmi** [ufal/neuralmonkey#210]
I have conducted several experiments to find out why the training is always killed after some time. I have found out that at the beginning of each validation something is loaded into the memory (I am using lazy mode on val_dataset) and then it is not released or reused.

When I disable validation, this problem do not occur (or maybe i didn't run the experiment for enough time to ocur).

I have tried to find the source of the problem but I do not see anything in the code which should stay in memory indefinitely

**Comment by jlibovicky** [comment]
Does it happen even if you don't use `LazyDataset` for the validation. My guess is it is because the validation file remains open after printing the examples in the log (because the file was not read until the end).

As a simple solution, I would just ban using lazy dataset for validation data. It makes sense to have them in memory because they are used often (unlike the training data).

**Comment by jlibovicky** [comment]
The proper solution would be: readers will use file handles instead of file names and a lazy dataset instance will keep a dictionary of the handles and every time you want to start reading the series, the dataset will check whether the series' file handle isn't open. We were also taking about asynchronous validation. In that case, we would need to keep the file-handle table for each thread separately.

**Comment by tomkocmi** [comment]
Yes, it does happen even when using Lazy mode. And I would guess it happens even faster. And also I am talking about 500MB per validation (when not using Lazy ... note that the validation set has onlu 200kb) so it will not be anything simple as forgotten pointer.

**Comment by jindrahelcl** [comment]
It sounds to me like a problem with CPU memory allocation in tensorflow.. Can we profile the code using something like this: https://pypi.python.org/pypi/memory_profiler ?

**Comment by tomasmcz** [comment]
Coudn't reproduce.

**Comment by tomkocmi** [comment]
I check the problem with monster decoder and it didn't appear.

**Comment by jindrahelcl** [comment]
There's no monster decoder in master for a week now - did you mean this master or you did not update the newest changes?

**Comment by tomkocmi** [comment]
I know there is no monster decoder in master :) ... I only wanted to specify, that I run it with the new version of decoder (master branch from last tuesday)
